### DISCUSSION TITLE
How to add Python packages to object-detector-python?

### DISCUSSION THREAD

**Discussion by igor-rud-kr** [AxisCommunications/acap-computer-vision-sdk-examples#43] [Q&A] [answered]
I tried adding the following lines to Dockerfile:

FROM python:3
RUN pip install --no-cache-dir --upgrade pip && pip install --no-cache-dir pytz requests paho-mqtt

It added about 500M to the image (not sure if there is a way to add just the packages I need), but then when I tried to start it using docker-compose, the following error message was displayed:

object-detector-python-object-detector-python-1  | standard_init_linux.go:219: exec user process caused: exec format error
object-detector-python-object-detector-python-1 exited with code 1

**Comment by Corallo** [comment, marked as answer]
Hello Igor, 

By adding `FROM python:3` ad the bottom of the file, you are building a docker image based on that, ignoring all the rest of the build. This will not work. You are going to need the libraries imported from the computer vison SDK.
Besides, the image that you are pulling, "python:3" is an x86 image, which is not the right architecture for axis cameras (that is either armv7hf or aarch64). 
If you need to add a python library you can simply do it without importing a new imaga, just like it is done in the [pose-estimation-example](https://github.com/AxisCommunications/acap-computer-vision-sdk-examples/blob/master/pose-estimator-with-flask/Dockerfile#L16).

**Comment by igor-rud-kr** [comment]
Sorry, one more question

I added the following line to the Dockerfile in object-detector-python project:
RUN pip install RUN pip install Flask

but when I try to build it, I still receive an error:

Step 13/16 : RUN pip install RUN pip install Flask
 ---> [Warning] The requested image's platform (linux/arm64/v8) does not match the detected host platform (linux/amd64) and no specific platform was requested
 ---> Running in 27e79ec59704
standard_init_linux.go:228: exec user process caused: exec format error
The command '/bin/sh -c pip install RUN pip install Flask' returned a non-zero code: 1

The Dockerfile seems to be identical with the one in pose-estimator-with-flask. The only difference that I see is the first line: ARG ARCH=aarch64 (in my Dockerfile) and ARG ARCH=armv7hf (in the pose-estimator-with-flask). Can it cause the error?

**Reply by Corallo** [reply]
Hello igor, first of all, that command seems wrong, if you want to add a python package just add
RUN pip install <package_name>.
Your error is indeed an architecture error, to use "RUN" commands on a different architecture you should first install qemu.
Running this command: `docker run -it --rm --privileged multiarch/qemu-user-static --credential yes --persistent yes`
you should be able to add RUN instructions to your dockerfile

**Comment by igor-rud-kr** [comment]
I think this happens because I am running build on linux/amd64 while requested platform is linux/arm64/v8

Not sure how to fix it. The step to install qemu to allow build flask for a different architecture produces a number of errors like this:
Setting /usr/bin/qemu-alpha-static as binfmt interpreter for alpha
sh: write error: Invalid argument

I also tried docker buildx, but it also fails because it does not pick up proxy settings

**Reply by igor-rud-kr** [reply]
UPDATE:

docker buildx build --platform linux/arm64/v8 --no-cache . -t acap4-object-detector-python --build-arg aarch64
still results in the same error:

 => ERROR [stage-3 6/8] RUN "pip install pytz requests paho-mqtt"                                                                                                                       0.3s
------
 > [stage-3 6/8] RUN "pip install pytz requests paho-mqtt":
#12 0.178 standard_init_linux.go:228: exec user process caused: exec format error
------
error: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c "pip install pytz requests paho-mqtt"]: exit code: 1

**Reply by Corallo** [reply]
Making qemu to work is the simplest way to solve this issue, not sure if buildx can substitute it.

**Reply by igor-rud-kr** [reply]
Thank you!

This link describe the same issue:
https://github.com/multiarch/qemu-user-static/issues/38

Will try to build it on a different machine

**Reply by Corallo** [reply]
Looks like that when you get that error it might be because qemu is already set.
Try to build again, or in alternative `docker run --rm --privileged multiarch/qemu-user-static:register --reset`.

Otherwise, probably is better to start again on a clean environment.

**Reply by igor-rud-kr** [reply]
Thank you so much! I was able to build acap4-object-detector-python on another machine using `docker run -it --rm --privileged multiarch/qemu-user-static --credential yes --persistent yes` command. On this machine the command completed successfully, and I was able to build this image with additional Python packages
